### PR TITLE
C3: ADC raw values calibration

### PR DESCRIPTION
**Pull request by JurajSadel**
This adds initial ADC raw values calibration. The solution is ported from https://github.com/apache/nuttx/blob/master/arch/risc-v/src/esp32c3/esp32c3_adc.c

The `11dB` attenuation seems to be working fine, however, other values(0, 2.5, and 6dB) are weird - I'm getting 4095(raw) for the default state (with `11dB` I'm getting ~2200(raw)). The weirdest thing here is that I have the same behavior in [ESP-IDF oneshot example](https://github.com/espressif/esp-idf/blob/master/examples/peripherals/adc/oneshot_read/main/oneshot_read_main.c). I've tested with Rust board (C3, revision3) and 5V joystick.

If anyone has any idea, what is going on with `attenuations` (or can test this code) it'd be very appreciated.
I'll look into calculating voltage from raw values in the meantime.

Related #326

### CODE REVIEW

**Comment by bjoernQ** [comment]
Tested on ESP32-C3 and I get the same values I get with the eps-idf example. (With `ADC_ATTEN_DB_0` and `ADC_ATTEN_DB_11`)

**Comment by bugadani** [comment]
How big an effort would be to extend this to the S3? Looking at the comment above the constants, it looks like this PR is based on S3 code. I can test 11dB, maybe 6dB attenuation in my project.

File: esp-hal-common/src/analog/adc/riscv.rs [hunk 1]
```diff
@@ -1,16 +1,71 @@
...
+#[cfg(not(esp32c6))]
+use crate::rom::rom_i2c_writeReg_Mask;
 use crate::{
     analog::ADC1,
     peripheral::PeripheralRef,
     peripherals::APB_SARADC,
     system::{Peripheral, PeripheralClockControl},
 };
 
+// constants taken from https://github.com/espressif/esp-idf/blob/045163a2ec99eb3cb7cc69e2763afd145156c4cf/components/soc/esp32s3/include/soc/regi2c_saradc.h
```
Comment (bugadani) [inline esp-hal-common/src/analog/adc/riscv.rs:22]: Please correct the link to point to the c3 header. At least `I2C_SAR_ADC_HOSTID` is different from the linked file

File: esp-hal-common/src/soc/esp32c3/efuse.rs [hunk 2]
```diff
@@ -59,4 +59,25 @@ impl Efuse {
         let efuse = unsafe { &*EFUSE::ptr() };
         efuse.rd_repeat_data1.read().wdt_delay_sel().bits()
     }
+
+    /// Read Efuse data and return them
+    pub fn read_efuse(address: u32, offset: u32, size: u32) -> u32 {
```
Comment (katyo) [inline esp-hal-common/src/soc/esp32c3/efuse.rs:64]: Efuses also can be read via peripheral access like so:
```rust
let efuse = unsafe { &*EFUSE::ptr() };

efuse.rd_sys_part1_data0.read().bit_field()
```
But seems currently bit fields isn't described.
(see registers definition in [esp32c3.base.svd](/esp-rs/esp-pacs/blob/4fe079193d4261319672bf91e4c11a5efb4a8268/esp32c3/svd/esp32c3.base.svd#L7542))

**Comment by elpiel** [comment]
Is there anything else I need to do apart from using the branch to test this out?

ADC values still don't show correct values:

Screenshot from 2023-06-12 14-02-41
I'm measuring a 250 mAh battery (4.2V max) with a voltage divider (2 x 470 k Ohms resistors) and these are the values I get.
With attenuation 11db I should see ~ (4.2V/2) = 2.1V for a full battery.

**Comment by JurajSadel** [comment]
@elpiel Could you try [this PR](https://github.com/esp-rs/esp-hal/pull/555), please?

**Comment by jessebraham** [comment]
Superseded by #555

**Pull request closed without merging**